lib/src/config_provider/schema.dart

```diff
@@ -0,0 +1,732 @@
+    resultCallback?.call(returnValue);
+    return returnValue;
+  }
+
+  /// Returns true if [value] is of Type [T].
+  bool checkType<T>({bool log = true}) {
+    if (value is! T) {
+      if (log) {
+        _logger.severe(
+            "Expected value of key '$pathString' to be of type '$T' (Got ${value.runtimeType}).");
+      }
+      return false;
+    }
+    return true;
+  }
+}
+
+class SchemaExtractionError extends Error {
+  final SchemaNode? item;
+  final String message;
+  SchemaExtractionError(this.item, [this.message = "Invalid Schema"]);
+
+  @override
+  String toString() {
+    if (item != null) {
+      return "$runtimeType: $message @ ${item!.pathString}";
+    }
+    return "$runtimeType: $message";
+  }
+}
+
+/// Base class for all Schemas to extend.
+abstract class Schema<E> {
+  /// Used to generate and refer the reference definition generated in json
+  /// schema. Must be unique for a nested Schema.
+  String? schemaDefName;
+
+  /// Used to generate the description field in json schema.
+  String? schemaDescription;
+
+  /// Custom validation hook, called post schema validation if successful.
+  bool Function(SchemaNode node)? customValidation;
+
+  /// Used to transform the payload to another type before passing to parent
+  /// nodes and [result].
+  dynamic Function(SchemaNode<E> node)? transform;
+
+  /// Called when final result is prepared via [extractNode].
+  void Function(SchemaNode<dynamic> node)? result;
+  Schema({
+    /// Used
+    required this.schemaDefName,
+    required this.schemaDescription,
+    required this.customValidation,
+    required this.transform,
+    required this.result,
+  });
+
+  bool validateNode(SchemaNode o, {bool log = true});
+
+  SchemaNode extractNode(SchemaNode o);
+
+  Map<String, dynamic> generateJsonSchemaNode(Map<String, dynamic> defs);
+
+  Map<String, dynamic> getJsonRefOrSchemaNode(Map<String, dynamic> defs) {
+    if (schemaDefName == null) {
+      return generateJsonSchemaNode(defs);
+    }
+    defs.putIfAbsent(schemaDefName!, () => generateJsonSchemaNode(defs));
+    return {r"$ref": "#/\$defs/$schemaDefName"};
+  }
+
+  Map<String, dynamic> generateJsonSchema(String schemaId) {
+    final defs = <String, dynamic>{};
+    final schemaMap = generateJsonSchemaNode(defs);
+    return {
+      r"$id": schemaId,
+      r"$schema": "https://json-schema.org/draft/2020-12/schema",
+      ...schemaMap,
+      r"$defs": defs,
+    };
+  }
+
+  /// Run validation on an object [value].
+  bool validate(dynamic value) {
+    return validateNode(SchemaNode(path: [], value: value));
+  }
+
+  /// Extract SchemaNode from [value]. This will call the [transform] for all
+  /// underlying Schemas if valid.
+  /// Should ideally only be called if [validate] returns True. Throws
+  /// [SchemaExtractionError] if any validation fails.
+  SchemaNode extract(dynamic value) {
+    return extractNode(SchemaNode(path: [], value: value));
+  }
+}
+
+class FixedMapKey<DE extends dynamic> {
```

Comment:
> Why extends dynamic instead of DE extends Object

We must use `Object?` (to support null). And that is the same as `dynamic` I suppose.


> Is it because the YAML and JSON abstractions also all use dynamic in the API so we don't have to write casts everywhere?

I initially did `FixedMapKey<DE>`, but the analyzer complained about specifying the generic type. I realized I had to just supply `<dynamic>` on most places. In general, the more nested configurations with different type of keys are, more `<dynamic>` is likely used in generic.

E.g, you can only do `FixedMapSchema<T>` if every value schema for each subkey, either returns T by itself or is converted to T using the `transform` closure.

Currently the Generic Type in the `Schema` is only used to enforce the type returned by calling extractNode, and also makes the value type-safe in the `transform` closure.